### PR TITLE
Ensure a split becomes mature after a given period since its creation.

### LINKED ISSUE #2033: Delete query does not affect immature splits and the merge policy can leave some merge immature forever.
By nature, immature splits should not remain immature. With a given merge policy, they should become mature, one day.
However, we don't know when they will become mature, and under certain assumptions, it may happen in a very long time. For example, if a merge policy has a huge `split_num_docs_target`... And in this case, we will never delete documents.

We can certainly assume that a user should know what he is doing but... delete queries are mainly here for GDPR reasons and it can be reassuring for a user to be able to easily check if his delete query has really deleted all documents everywhere or ensure that we delete documents even on immature splits.

At first, a clear documentation is needed, and according to feedback on this feature, we may fix (or not) this issue.

### PR DESCRIPTION
**Pull request by fmassot**
### Description

Fix #2033 

I added a new maturity condition: 
- if a split is created since a given `maturity_period`, it will be considered mature. I did not use the `publish_timestamp` more for conveniency as it's not present in the split metadata and instead used the `create_timestamp` which is almost the same (a few minutes difference at maximum). For the maturity condition, I believe this is fine.

One negative note: if someone adds a new merge policy, there is no mechanism to warn him that a split can never become mature. We might want to add a notion of maturity period at the merge policy trait level and add a test that can run on any merge policy to check that a split really becomes mature after this maturity period. 

I will follow-up with another PR to make use of publish_timestamp filtering when we retrieve splits merge candidates and stale splits.

### CODE REVIEW

File: quickwit/quickwit-config/src/index_config/mod.rs [hunk 1]
```diff
@@ -758,6 +758,7 @@ mod tests {
             MergePolicyConfig::StableLog(StableLogMergePolicyConfig {
                 merge_factor: 9,
                 max_merge_factor: 11,
+                maturity_period: Duration::from_secs(48 * 3600),
```
Comment (fmassot) [inline quickwit/quickwit-config/src/index_config/mod.rs:761]: I found the `24h` period a bit aggressive and ended up putting 48h.

File: quickwit/quickwit-config/src/merge_policy_config.rs [hunk 2]
```diff
@@ -30,6 +33,11 @@ pub struct ConstWriteAmplificationMergePolicyConfig {
     /// Maximum number of merges that a given split should undergo.
     #[serde(default = "default_max_merge_ops")]
     pub max_merge_ops: usize,
+    /// Maximum number of merges that a given split should undergo.
```
Comment (fmassot) [inline quickwit/quickwit-config/src/merge_policy_config.rs:36]: I need to update the docs.

File: quickwit/quickwit-config/src/merge_policy_config.rs [hunk 3]
```diff
@@ -53,6 +63,11 @@ pub struct StableLogMergePolicyConfig {
     /// Maximum number of splits that can be merged together in a single merge operation.
     #[serde(default = "default_max_merge_factor")]
     pub max_merge_factor: usize,
+    /// Maximum number of merges that a given split should undergo.
```
Comment (fmassot) [inline quickwit/quickwit-config/src/merge_policy_config.rs:66]: Update the docs.

File: quickwit/quickwit-config/src/merge_policy_config.rs [hunk 4]
```diff
@@ -71,18 +92,42 @@ fn default_min_level_num_docs() -> usize {
...
+
+fn serialize_duration<S>(value: &Duration, s: S) -> Result<S::Ok, S::Error>
+where S: Serializer {
+    let value_str = humantime::format_duration(*value).to_string();
+    s.serialize_str(&value_str)
+}
+
 #[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
 #[serde(tag = "type")]
+#[serde(deny_unknown_fields)]
```
Comment (fmassot) [inline quickwit/quickwit-config/src/merge_policy_config.rs:130]: Always better to deny.

File: quickwit/quickwit-config/src/index_config/mod.rs [hunk 5]
```diff
@@ -856,6 +857,30 @@ mod tests {
         assert_eq!(minimal_config.doc_mapping.mode, ModeType::Lenient);
     }
 
+    #[test]
+    fn test_index_config_with_malformated_maturity_duration() {
```
Comment (guilload) [inline quickwit/quickwit-config/src/index_config/mod.rs:861]:
```suggestion
    fn test_index_config_with_malformed_maturity_duration() {
```

File: quickwit/quickwit-config/resources/tests/index_config/hdfs-logs.json [hunk 6]
```diff
@@ -54,7 +54,8 @@
         "merge_policy": {
             "type": "stable_log",
             "merge_factor": 9,
-            "max_merge_factor": 11
+            "max_merge_factor": 11,
+            "maturity_period": "48 hours"
```
Comment (guilload) [inline quickwit/quickwit-config/resources/tests/index_config/hdfs-logs.json:58]: Technically, that's an immaturity period. `maturation_period` might be what we are looking for.

File: quickwit/quickwit-config/src/merge_policy_config.rs [hunk 7]
```diff
@@ -30,6 +33,14 @@ pub struct ConstWriteAmplificationMergePolicyConfig {
     /// Maximum number of merges that a given split should undergo.
     #[serde(default = "default_max_merge_ops")]
     pub max_merge_ops: usize,
+    /// Duration relative to `split.created_timestamp` after which a split
+    /// becomes mature.
+    /// If `now() >= split.created_timestamp + maturity_period` then
+    /// the split is mature.
```
Comment (guilload) [inline quickwit/quickwit-config/src/merge_policy_config.rs:39]:
```suggestion
    /// the split is considered mature.
```

File: quickwit/quickwit-config/src/merge_policy_config.rs [hunk 8]
```diff
@@ -71,18 +92,42 @@ fn default_min_level_num_docs() -> usize {
...
         StableLogMergePolicyConfig {
             min_level_num_docs: default_min_level_num_docs(),
             merge_factor: default_merge_factor(),
             max_merge_factor: default_max_merge_factor(),
+            maturity_period: default_maturity_period(),
         }
     }
 }
 
+fn parse_duration<'de, D>(deserializer: D) -> Result<Duration, D::Error>
```
Comment (guilload) [inline quickwit/quickwit-config/src/merge_policy_config.rs:110]:
```suggestion
fn parse_human_duration<'de, D>(deserializer: D) -> Result<Duration, D::Error>
```

File: quickwit/quickwit-config/src/merge_policy_config.rs [hunk 9]
```diff
@@ -71,18 +92,42 @@ fn default_min_level_num_docs() -> usize {
...
+where D: Deserializer<'de> {
+    let value: String = Deserialize::deserialize(deserializer)?;
+    let duration = humantime::parse_duration(&value).map_err(|error| {
+        de::Error::custom(format!(
+            "Cannot parse human duration `{}`: {:?}",
```
Comment (guilload) [inline quickwit/quickwit-config/src/merge_policy_config.rs:115]:
```suggestion
            "Failed to parse human-readable duration `{value}`: {error:?}",
```

File: quickwit/quickwit-indexing/src/merge_policy/nop_merge_policy.rs [hunk 10]
```diff
@@ -56,7 +56,7 @@ mod tests {
             .into_iter()
             .next()
             .unwrap();
-        // All splits are mature when merge is disabled.
+        // All splits are mature always mature for `NopMergePolicy`.
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/merge_policy/nop_merge_policy.rs:59]: // All the splits are always mature for the `NopMergePolicy`.

File: quickwit/quickwit-indexing/src/merge_policy/stable_log_merge_policy.rs [hunk 11]
```diff
@@ -116,8 +117,17 @@ impl MergePolicy for StableLogMergePolicy {
         operations
     }
 
+    /// A mature split for merge is a split that won't undergo merge operation in the future.
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/merge_policy/stable_log_merge_policy.rs:120]:
```suggestion
    /// A mature split for merge is a split that won't undergo any merge operation in the future.
```

**Review by guilload — APPROVED** [review]

**Pull request merged**